src/tests/suite/dfuse_test.c

```diff
@@ -176,25 +189,154 @@ do_ioctl(void **state)
 	assert_int_equal(dur.uid, geteuid());
 	assert_int_equal(dur.gid, getegid());
 
+	rc = unlinkat(root, "ioctl_file", 0);
+	assert_return_code(rc, errno);
+
 	rc = close(root);
 	assert_return_code(rc, errno);
 }
 
-int
-main(int argc, char **argv)
+static bool
+timespec_gt(struct timespec t1, struct timespec t2)
+{
+	if (t1.tv_sec == t2.tv_sec)
+		return t1.tv_nsec > t2.tv_nsec;
+	else
+		return t1.tv_sec > t2.tv_sec;
+}
+
+void
+do_mtime(void **state)
+{
+	struct stat     stbuf;
+	struct timespec prev_ts;
+	struct timespec now;
+	struct timespec times[2];
+	int             fd;
+	int             rc;
+	char            input_buf[] = "hello";
+	int             root        = open(test_dir, O_PATH | O_DIRECTORY);
+
+	assert_return_code(root, errno);
+
+	/* Open a file and sanity check the mtime */
+	fd = openat(root, "my_file", O_RDWR | O_CREAT | O_EXCL, S_IWUSR | S_IRUSR);
+	assert_return_code(fd, errno);
+	rc = clock_gettime(CLOCK_REALTIME, &now);
+	assert_return_code(fd, errno);
+	rc = fstat(fd, &stbuf);
+	assert_return_code(rc, errno);
+	prev_ts.tv_sec  = stbuf.st_mtim.tv_sec;
+	prev_ts.tv_nsec = stbuf.st_mtim.tv_nsec;
+	assert_true(now.tv_sec - prev_ts.tv_sec < 3);
+
+	/* Write to the file and verify mtime is newer */
+	rc = write(fd, input_buf, sizeof(input_buf));
+	assert_return_code(rc, errno);
+	rc = fstat(fd, &stbuf);
+	assert_return_code(rc, errno);
+	assert_true(timespec_gt(stbuf.st_mtim, prev_ts));
+	prev_ts.tv_sec  = stbuf.st_mtim.tv_sec;
+	prev_ts.tv_nsec = stbuf.st_mtim.tv_nsec;
+
+	/* Truncate the file and verify mtime is newer */
+	rc = ftruncate(fd, 0);
+	assert_return_code(rc, errno);
+	rc = fstat(fd, &stbuf);
+	assert_return_code(rc, errno);
+	assert_true(timespec_gt(stbuf.st_mtim, prev_ts));
+	prev_ts.tv_sec  = stbuf.st_mtim.tv_sec;
+	prev_ts.tv_nsec = stbuf.st_mtim.tv_nsec;
+
+	/* Set and verify mtime set in the past */
+	times[0]         = now;
+	times[1].tv_sec  = now.tv_sec - 10;
+	times[1].tv_nsec = 20;
+	rc               = futimens(fd, times);
+	assert_return_code(fd, errno);
+	rc = fstat(fd, &stbuf);
+	assert_return_code(rc, errno);
+	assert_int_equal(stbuf.st_mtim.tv_sec, times[1].tv_sec);
+	assert_int_equal(stbuf.st_mtim.tv_nsec, times[1].tv_nsec);
+	prev_ts.tv_sec  = stbuf.st_mtim.tv_sec;
+	prev_ts.tv_nsec = stbuf.st_mtim.tv_nsec;
+
```

Comment:
after manually setting the mtime in the previous step, it would be good to check that updating the file then stat, actually returns the more recent mtime.

Reply:
This does that does't it?  Do you mean an extra check where we write to the file, then stat again and ensure that the file is newer than prev_ts?

Reply:
yes indeed. here we set mtime with futimens. and verify that it's the one we set.
we just want to make sure if we touch the file again, that one we set is overwritten.
(just to verify an internal implementation detail). i did that in my dfs unit test, so it's not necessary to do here, but i see this test is mimicking the dfs one, so maybe we can just add it to make sure the dfuse level test behaves the same